### ISSUE TITLE
Parallelization of `kfold()` (across CV folds) on Windows

### ISSUE THREAD

**Issue by fweber144** [stan-dev/rstanarm#551]
#### Summary:
On Windows, the parallelization of `kfold()` across the CV folds doesn't always work (throws an error).

#### Description:
The issue mentioned in "Summary" occurs if the `stan_<...>()` call uses objects in some arguments (see "Reproducible Steps" below). Perhaps the reason is that there are no exports after the following lines?: https://github.com/stan-dev/rstanarm/blob/45e0707b0ab8c8e16d65e7f2af70a98bfaa93363/R/loo-kfold.R#L226-L227

#### Reproducible Steps:
```r
data("df_gaussian", package = "projpred")
dat_gauss <- data.frame(y = df_gaussian$y, df_gaussian$x)
D <- sum(grepl("^X", names(dat_gauss)))
p0 <- 5
N <- nrow(dat_gauss)
( tau0 <- p0 / (D - p0) * 1 / sqrt(N) )
library(rstanarm)
options(mc.cores = parallel::detectCores(logical = FALSE)) # gives 4 on my machine
rfit <- stan_glm(y ~ .,
                 data = dat_gauss,
                 prior = hs(global_scale = tau0),
                 QR = TRUE,
                 seed = 1669262042)
```
Now inner parallelization via `kfold.stanreg()`'s internal object `stan_cores` works (doesn't throw an error) and gives the (correct) messages `Fitting model 1 out of 4`, ..., `Fitting model 4 out of 4`:
```r
rkfold_one_kfold_core <- kfold(rfit, K = 4, cores = 1)
```

However, outer parallelization across CV folds via `kfold.stanreg()`'s internal object `kfold_cores` doesn't work:
```r
rkfold <- kfold(rfit, K = 4)
```
That last line gives the (correct) message `Fitting K = 4 models distributed over 4 cores`, but throws the error
```
Error in checkForRemoteErrors(val) :
  4 nodes produced errors; first error: object 'tau0' not found
```

#### RStanARM Version:
2.21.2 (from <https://mc-stan.org/r-packages/>)

#### R Version:
4.1.1

#### Operating System:
Windows 10 x64

**Comment by jgabry** [comment]
Following up on this based on https://discourse.mc-stan.org/t/projpred-error/37314/3. Sorry I never saw this previously. As a non-Windows user I'm not quite sure how to fix this. It seems like the issue here is that `tau0` would need to be included in `clusterExport()`, but `tau0` is just in the global environment. How would we know which variables in the global environment we need to export (we wouldn't want to do everything since there could be large irrelevant objects in the global environment)?

**Comment by fweber144** [comment]
Yes, this is definitely an issue. The [**future**](https://doi.org/10.32614/CRAN.package.future) package handles such exports automatically. In **projpred**, we use the [**foreach**](https://doi.org/10.32614/CRAN.package.foreach) package for parallelization (where necessary, we use it in combination [**doRNG**](https://doi.org/10.32614/CRAN.package.doRNG)), e.g., in lines https://github.com/stan-dev/projpred/blob/e15204de5791fa9691f4f402af29d8d0ff90a497/R/cv_varsel.R#L1317-L1335 or https://github.com/stan-dev/projpred/blob/e15204de5791fa9691f4f402af29d8d0ff90a497/R/divergence_minimizers.R#L87-L117. The **future** package (and hence its automatic exports) can then be used via the [**doFuture**](https://doi.org/10.32614/CRAN.package.doFuture) backend for **foreach**. Illustrations can be found in the tests (the relevant lines are https://github.com/stan-dev/projpred/blob/e15204de5791fa9691f4f402af29d8d0ff90a497/tests/testthat/test_parallel.R#L8-L25) and in the [main vignette](https://mc-stan.org/projpred/articles/projpred.html#final-cv_varsel-run) (but note that the **future** package is not used in the main vignette). I hope this helps. And thanks for following up on this!
